animation_nodes/algorithms/mesh/triangulate_mesh.pyx

```diff
@@ -0,0 +1,316 @@
+    cdef Py_ssize_t previousIndex, nextIndex
+    previousIndex = verticesData[earIndex].previous
+    nextIndex = verticesData[earIndex].next
+
+    verticesData[earIndex].isEar = False
+    verticesData[nextIndex].previous = previousIndex
+    verticesData[previousIndex].next = nextIndex
+
+    cdef float angle
+    angle = calculateAngleWithSign(verticesData, previousIndex)
+    verticesData[previousIndex].angle = angle
+    if angle >= 0.0 and not pointsInTriangle(polyLength, verticesData, previousIndex):
+        verticesData[previousIndex].isEar = True
+    else:
+        verticesData[previousIndex].isEar = False
+
+    angle = calculateAngleWithSign(verticesData, nextIndex)
+    verticesData[nextIndex].angle = angle
+    if angle >= 0.0 and not pointsInTriangle(polyLength, verticesData, nextIndex):
+        verticesData[nextIndex].isEar = True
+    else:
+        verticesData[nextIndex].isEar = False
+
+cdef struct Vertex:
+    Py_ssize_t index
+    Vector3 location
+    float angle
+    bint isEar
+    Py_ssize_t previous
+    Py_ssize_t next
+
+# Make sure normal (+z) is correct.
+@cython.cdivision(True)
+cdef bint polyCounterClockwise(Vertex* verticesData):
+    cdef Vector3 v1, v2
+    cdef float area = 0.0
+    cdef Py_ssize_t headIndex = 0
+    cdef Py_ssize_t currentIndex
+
+    currentIndex = headIndex
+    while(True):
+        v1 = verticesData[currentIndex].location
+        currentIndex = verticesData[currentIndex].next
+        v2 = verticesData[currentIndex].location
+        area += (v2.x - v1.x) * (v2.y + v1.y)
+        if currentIndex == headIndex: break
+
+    if area > 0.0: return False
+    return True
+
+# Calculate inner angle.
+cdef float calculateAngleWithSign(Vertex* verticesData, Py_ssize_t index):
+    cdef Vector3 v1 = verticesData[verticesData[index].previous].location
+    cdef Vector3 v2 = verticesData[index].location
+    cdef Vector3 v3 = verticesData[verticesData[index].next].location
+    cdef Vector3 ab, bc
+    cdef float angle
+
+    subVec3(&ab, &v2, &v1)
+    subVec3(&bc, &v2, &v3)
+
+    normalizeVec3_InPlace(&ab)
+    normalizeVec3_InPlace(&bc)
+
+    angle = angleNormalizedVec3(&ab, &bc)
+
+    # Make sure normal (+z) is correct, and the polygon is counter clockwise. Convex vertex has sign >= 0.
+    cdef float sign = (v2.x - v1.x) * (v3.y - v1.y) - (v2.y - v1.y) * (v3.x - v1.x)
+    if sign < 0.0: return -(angle - angle % 0.001)
+    return (angle - angle % 0.001)
+
+# Find the index of minimum angle.
+cdef int findEarHasMinAngle(Py_ssize_t polyLength, Vertex* verticesData):
+    cdef float angleMin = 3.14
+    cdef float angle
+    cdef Py_ssize_t i, angleMinIndex
+
+    angleMinIndex = 0
+    for i in range(polyLength):
+        if verticesData[i].isEar:
+            angle = abs(verticesData[i].angle)
+            if angleMin > angle:
+                angleMin = angle
+                angleMinIndex = i
```

Comment:
You are not looping over the linked list correctly. More on this later.

Reply:
It is changed.